pyro/infer/mcmc/mh.py

```diff
@@ -0,0 +1,255 @@
+        self._tune_cnt = self.tune_frequency
+
+    def sample(self, site_node):
+        cur_val = site_node['value']
+        mu = self.mu.expand_as(cur_val)
+        sigma = self.scale*self.sigma
+
+        # sigma just a number? expand to equal the must
+        if tuple(sigma.data.shape) != mu.data.shape:
+            sigma = sigma.expand_as(mu.data)
+
+        return cur_val + dist.Normal(mu, sigma).sample()
+
+    def log_pdf_given(self, y, given_x):
+        # center our distribution at mu = given_x, then
+        # get log pdf of y
+        mu = given_x
+        sigma = self.scale*self.sigma
+        return dist.Normal(mu, sigma).log_pdf(y)
+
+    # borrowed from:
+    # https://github.com/mcleonard/sampyl/blob/master/sampyl/samplers/metropolis.py#L102
+    # which in turn borrows from pymc3
+    def tune(self, acceptance):
+        self._tune_cnt -= 1
+        if self._tune_cnt > 0:
+            return self
+
+        scale = self.scale
+        # Switch statement
+        if acceptance < 0.001:
+            # reduce by 90 percent
+            scale *= 0.1
+        elif acceptance < 0.05:
+            # reduce by 50 percent
+            scale *= 0.5
+        elif acceptance < 0.2:
+            # reduce by ten percent
+            scale *= 0.9
+        elif acceptance > 0.95:
+            # increase by factor of ten
+            scale *= 10.0
+        elif acceptance > 0.75:
+            # increase by double
+            scale *= 2.0
+        elif acceptance > 0.5:
+            # increase by ten percent
+            scale *= 1.1
+
+        self.scale = scale
+        # reset our tuning count
+        self._tune_cnt = self.tune_frequency
+        return self
+
+
+class MH(TraceKernel):
+    def __init__(self, model, proposal_dist):
+        self.model = model
+        self.proposal_dist = proposal_dist
+
+        # store all of our traces, and keep count on acceptance ratio
+        self._kernel_traces = []
+        self._accept_cnt = None
+        self._tune_cnt = None
+        self._call_cnt = None
+
+        super(MH, self).__init__()
+
+    def setup(self, *args, **kwargs):
+
+        # haven't accepted any!
+        self._accept_cnt = 0
+        self._tune_cnt = 0
+        self._call_cnt = 0
+
+        # store our args call for later?
+        self._args = args
+        self._kwargs = kwargs
+
+    def initial_trace(self):
+        # maintain all traces, may need to kill at the end
+        trace = poutine.trace(self.model).get_trace(*self._args, **self._kwargs)
+        assert_valid_trace(trace)
+        self._kernel_traces.append(trace)
+        return trace
+
+    def cleanup(self):
+        for trace in self._kernel_traces:
+            self._cleanup_trace(trace)
+        self._kernel_traces = []
+
+    def _cleanup_trace(self, reject_trace, accept_trace=None):
+        pass
+
+    @property
+    def num_accepts(self):
+        return self._accept_cnt
+
+    def _get_trace_proposal(self, trace, r_site, proposal_value):
```

Comment:
This method is a nice conceptual unit that we could reuse. I'd recommend factoring it out into a free standing helper either in this file or in poutine/trace.py:
```py
def truncate_trace(tr, end_site):
    """
    Truncate a trace up until (but not including) an ``end_site``.
    :param Trace tr: A source trace.
    :param str end_site: The name of an end site to truncate after.
    """
    ...your logic here...
```